internal/event/dispatch_service.go

```diff
@@ -84,7 +63,6 @@ func (s *DispatchService) Stop() {
 	if !atomic.CompareAndSwapInt32(&s.state, ready, stopped) {
 		return
 	}
-	close(s.doneCh)
```

Comment:
I think the stop method of a dispath service should guarantee that all events are handled before stopping.

Reply:
For sync events that's guaranteed. For async events, that would require an extra mechanism, like passing a channel to the handler which would be very error-prone to use.

Reply:
When we don't guarantee that all events are consumed, would that cause any correctness problems where it is used ?

Reply:
In the current design, events are not buffered, so we can assume they are consumed. Even if they were buffered, `Stop` of the dispatcher would only be called during client shutdown, so it should be safe to assume that the remaining events are not useful (and probably shouldn't even be delivered to the user).

Reply:
We wait for events to be handled in the java, otherwise we consider it as a resource leak. 
If you open and close a client fast enough, the system can go out of memory. 

Imagine a user I puts a long running task on the Shutdown method . Opens a client, put an item, and shutdown the client per request. And it has very high throughput. 
Result: The system does not slows down as it should be but it goes out of memory eventually crashing the whole service.

Reply:
https://github.com/hazelcast/hazelcast-go-client/pull/644#discussion_r711845847

I am confused about this comment of yours. Let me ask the following. 

When hazelcastClient.shutdown() finished. 

Is it guaranteed that handling LifecycleServiceShutdown is started ? And also, 
is it guaranteed that handling LifecycleServiceShutdown is finished ?

Reply:
Ok. If for the async subscribe if it is not guaranteed, lets put it on the Subscribe doc. Calling this method does not guarantee that this event will be delivered to handlers when dispatch service is stopped from another thread.

Reply:
In the current design (this PR), we don't use channels, and the corresponding handlers are guaranteed to be started before starting the next handler and once started, handlers run to completion. I would say events are guaranteed to be delivered to handlers, even if the dispatch service is stopped.

Reply:
Firstly I need to correct my wording on the https://github.com/hazelcast/hazelcast-go-client/pull/644#discussion_r711962431

I wrote :
Calling this method does not guarantee that this event will be delivered to handlers when dispatch service is stopped from **another** thread.
This is true because a second thread can set `state` to stopped and the callers events may not be fired. Right ? 
 And callers of the Publish is not aware of it. So they need to know at least from doc.
 Or we can return a boolean to denote if the publish will be handled eventually or not. 

But actually, I meant to say :
Calling this method does not guarantee that this event will be **handled(and finished running)** when dispatch service is stopped from **same** goroutine. Meaning: 
```
s.Publish(event)
s.Close()
```
When close is finished, we don't know that the handler is finished handling the event. This is not guaranteed. 
We can document this as well clearly on the Subscribe.

Reply:
In the current design, stopping the dispatch service only disables publishing/subscribing/unsubscribing an event. Once an event is published, all handlers for that event run in a goroutine, even if the dispatch service is stopped, and stopping the dispatch service does not "kill" those goroutines. If a handler goroutine is running when the client is shutdown, it will continue to run, until the process itself is killed.

Reply:
Yes, I got how it works. I am just saying documentation of the method can state this so that users of it will not make mistake.